### PR TITLE
task/SWTCH-1983 restrict cross-request origins

### PR DESCRIPTION
**Pull request by JGiter**
List of changes
1. Adds options to restrict origin of login request. When `RESTRICT_CORS_ORIGIN` environment variable is set to `true` only origins defined in `ALLOWED_ORIGINS` environment variable will be able to login. `ALLOWED_ORIGINS` are specified as comma separated values
2. Adds check that matches request origin to origin of `refresh_token`

If login request has no origin, then it is authenticates regardless of `RESTRICT_CORS_ORIGIN`

### CODE REVIEW

File: src/db/cli.ts
Comment (JGiter) [inline src/db/cli.ts:15]: Attempt to fix https://github.com/energywebfoundation/ssi-hub/actions/runs/4066246957/jobs/7002070873#step:11:161

File: e2e/claim/claim-with-request.testSuite.ts
Comment (JGiter) [inline e2e/claim/claim-with-request.testSuite.ts:318]: removed this bc invalid origin will not be able reach RoleClaim constructor now since origin is checked during authentication

File: src/db/cli.ts
Comment (artursudnik) [reply src/db/cli.ts]: is this still necessary?

Comment (artursudnik) [reply src/db/cli.ts]: oh, nevermind. it's outdated :)

Comment (JGiter) [reply src/db/cli.ts]: sorry, forget to rm comment. will rm now

File: src/modules/auth/token.service.ts
Comment (jrhender) [inline src/modules/auth/token.service.ts:128]: As I think this is a bit separate from the direct CORS configuration and is a refactoring rather than new functionality, I think it would be better to have this in a separate PR. We can leave it here, but just a point to keep in mind for future.

**Review by jrhender — CHANGES_REQUESTED** [review]
Nice work on the test cases 👍 

I think the most important change is that I think it would be better to use the `cors` package (via Nest config) rather than rolling our own origin check.

**Comment by JGiter** [comment]
> I think the most important change is that I think it would be better to use the `cors` package (via Nest config) rather than rolling our own origin check.

I added custom middleware because from the first look at cors configuration options I didn't see a way to pass requests without `Origin` header, but now I see that it can be done dynamically. I'll try this

File: src/env-vars-validation-schema.ts
Comment (JGiter) [reply src/env-vars-validation-schema.ts:110]: I tried to verify `ALLOWED_ORIGINS` as array, but it failed. I assume that all values in `.env` are strings. I can't cite authoritative source for this, but here related discussion https://stackoverflow.com/questions/31552125/defining-an-array-as-an-environment-variable-in-node-js. The point of interest is the second answer

File: e2e/auth/login.testSuite.ts
Comment (JGiter) [reply e2e/auth/login.testSuite.ts:120]: We have one test when origin header isn't set https://github.com/energywebfoundation/ssi-hub/blob/44b16b3694f27773b0ea9ba02b26e6fe295ba657/e2e/auth/login.testSuite.ts#L35. But I agree that some more can be added. Grouping tests when origin is set or not will also help.
By the way from my opinion request to `v1/search/test` in your test should fail, because `Origin` header doesn't match origin of access token, which undefined. Why we should allow this request?

File: src/modules/auth/auth.module.ts
Comment (JGiter) [reply src/modules/auth/auth.module.ts]: I agree. Though `app.enableCors` is more consistent with how we adding other middleware, then `NestFactory.create`. We are calling methods on `app`

File: src/modules/auth/token.service.ts
Comment (JGiter) [reply src/modules/auth/token.service.ts:128]: Yes, it is not directly relates to CORS. I separated matching origin to token in order not to repeat it separately for access and refresh token

File: e2e/auth/login.testSuite.ts
Comment (jrhender) [reply e2e/auth/login.testSuite.ts:161]: Ah, thanks, I didn't see the test that you linked to. I think that test is probably enough. In fact, the test you linked to is better because it is retrieving the token from the cookie header in the response and not from the response body (see https://energyweb.atlassian.net/browse/SWTCH-2020)

File: src/modules/auth/auth.module.ts
Comment (jrhender) [reply src/modules/auth/auth.module.ts]: Ah okay, I didn't realize that `app.enableCors` and the `cors` object in `NestFactory.create` are the same thing but that makes sense. You're right, it's better to modify the line that already has `app.enableCors` rather than switch the pattern we're using https://github.com/energywebfoundation/ssi-hub/blob/605820d672392b0ee72c107c2d150aef39d9b4c1/src/main.ts#L73

File: src/env-vars-validation-schema.ts
Comment (jrhender) [reply src/env-vars-validation-schema.ts:110]: Ah makes sense, thanks for the explanation 👍

**Review by jrhender — COMMENTED** [review]
@JGiter how hard would it be to add an e2e test for the new "allow all origin" case? Is it hard to have an e2e test with different config?

File: src/common/test.utils.ts
Comment (jrhender) [inline src/common/test.utils.ts:49]: It occurred to me that this code is duplicating the code in `main.ts`. Maybe we can do it in a separate PR, but is there a way to remove this duplication? Do you think this is worth doing?

**Comment by JGiter** [comment]
> @JGiter how hard would it be to add a test for the new "allow all origin" case? Is it hard to have a test with different config?

Unfortunately I don't see a simple way to change environment variable `RESTRICT_CROSS_ORIGINS` per test. There are two challenges:
- in order to change CORS configuration application has to be relaunched. So we have to start/stop application in each test, which results in longer CI
- CORS configuration have to be done in top level `describe`, where application is started, but inner `describe` has no way to hook into top-level
Nevertheless I added test for login with reflected origin, which is skipped by default (when `RESTRICT_CROSS_ORIGINS` is true), but can be enabled when all origins allowed https://github.com/energywebfoundation/ssi-hub/blob/1c42180a15d8331a6e76cb25499a8965cdf8becb/e2e/auth/login.testSuite.ts#L159

File: src/common/test.utils.ts
Comment (JGiter) [reply src/common/test.utils.ts:49]: Yes, this duplication should be removed

**Review by jrhender — COMMENTED** [review]
It looks good, thanks @JGiter . Before merging, let's coordinate with the DDHub team so that they are ready to test when we merge.

**Comment by jrhender** [comment]
Though the PR looks good and I think could be merged, I will add the "Do Not Merge" while we coordinate the deployment of this change with the DDHub change and the wider AEMO community.

@JGiter To help with communicating this change, could you add a description to this PR please?

File: devops/staging/values.yaml
Comment (jrhender) [inline devops/staging/values.yaml:121]: As we now have the deployment for gp4btc elsewhere, the only ui that we need access the "staging" environment is switchboard-staging
```suggestion
    RESTRICT_CORS_ORIGINS: true
    ALLOWED_ORIGINS: https://switchboard-staging.energyweb.org
```

**Review by jrhender — COMMENTED** [review]
A small comment about adjusting `staging` values now that we have separate gp4btc environment

**Comment by jrhender** [comment]
@JGiter FYI, I added prod values (https://github.com/energywebfoundation/ssi-hub/pull/562/commits/d47474982e895f15da6ec36f6b0e23571771e5ff) so that we could potentially merge https://github.com/energywebfoundation/ssi-hub/pull/576

**Review by jrhender — APPROVED** [review]

**Pull request merged**
